File: src/transform2d.c
/*
 * 2D Transform PAL chroma separation.
 *
 * Port of ld-chroma-decoder's TransformPal2D, itself based on Jim
 * Easterbrook's pyctools-pal (http://www.jim-easterbrook.me.uk/pal/):
 * windowed overlapping tiles are moved to the frequency domain, where
 * bins are kept only if they are symmetric with their reflection about
 * the chroma carrier, then overlap-added back. Runs in single-precision
 * float; everything downstream is fixed point.
 */

#include <math.h>
#include <pthread.h>
#include <string.h>

#include "transform2d.h"

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

#define XTILE COMP_T2D_XTILE
#define YTILE COMP_T2D_YTILE
#define XCOMPLEX COMP_T2D_XCOMPLEX
#define YCOMPLEX COMP_T2D_YCOMPLEX

/* level padding tiles see beyond the raster edges: black, 0x4000 (the
 * raster carries no blanking/burst samples, unlike a full stored line) */
#define LEVEL_BLACK 16384.0f

/* FFTW's planner is not thread-safe; execute-with-new-arrays is */
static pthread_mutex_t planner_lock = PTHREAD_MUTEX_INITIALIZER;

static double compute_window(int element, int limit)
{
    return 0.5 - 0.5 * cos((2.0 * M_PI * (element + 0.5)) / limit);
}

int comp_transform2d_init(comp_transform2d_t *t, double threshold, int level,
                          double evidence)
{
    if (!(threshold > 0.0 && threshold <= 1.0) || evidence < 0.0)
        return -1;
    t->level = !!level;
    t->use_lut = 0;
    t->evidence = (float)evidence;
    for (int i = 0; i < COMP_T2D_NTHRESH; i++)
        t->threshold_sq[i] = (float)(threshold * threshold);

    for (int y = 0; y < YTILE; y++)
        for (int x = 0; x < XTILE; x++)
            t->window[y][x] = (float)(compute_window(y, YTILE) * compute_window(x, XTILE));

    /* FFTW_UNALIGNED lets per-call tile buffers live on the stack, which
     * keeps frame processing allocation-free under fmParallel */
    float real[YTILE * XTILE];
    fftwf_complex cplx[YCOMPLEX * XCOMPLEX];
    pthread_mutex_lock(&planner_lock);
    t->forward = fftwf_plan_dft_r2c_2d(YTILE, XTILE, real, cplx,
                                       FFTW_ESTIMATE | FFTW_UNALIGNED);
    t->inverse = fftwf_plan_dft_c2r_2d(YTILE, XTILE, cplx, real,
                                       FFTW_ESTIMATE | FFTW_UNALIGNED);
    pthread_mutex_unlock(&planner_lock);
    if (!t->forward || !t->inverse) {
        comp_transform2d_free(t);
        return -1;
    }
    return 0;
}

void comp_transform2d_set_lut(comp_transform2d_t *t, const double *v)
{
    for (int b = 0; b < COMP_T2D_NTHRESH; b++)
        for (int k = 0; k < COMP_LUT_K; k++)
            t->lut[b][k] = (float)v[b * COMP_LUT_K + k];
    t->use_lut = 1;
}

void comp_transform2d_free(comp_transform2d_t *t)
{
    pthread_mutex_lock(&planner_lock);
    if (t->forward)
        fftwf_destroy_plan(t->forward);
    if (t->inverse)
        fftwf_destroy_plan(t->inverse);
    pthread_mutex_unlock(&planner_lock);
    t->forward = NULL;
    t->inverse = NULL;
}

/* gain from a per-bin LUT row: linear interpolation over the
 * pair-symmetry ratio, knots uniform on [0, 1] */
static inline float lut_gain(const float *row, float lo, float hi)
{
    const float r = hi > 0.0f ? lo / hi : 1.0f;
    const float pos = r * (COMP_LUT_K - 1);
    int k = (int)pos;
    if (k > COMP_LUT_K - 2)
        k = COMP_LUT_K - 2;
    return row[k] + (row[k + 1] - row[k]) * (pos - k);
}

/* Keep only bins that look like chroma: modulated chroma is symmetric
 * about the carrier at (fsc, 72 c/aph) = (XTILE/4, YTILE/4). In
 * threshold mode, compare each candidate bin's squared magnitude with
 * its reflection and keep the pair only if they match within the
 * threshold ratio. In level mode, set the larger of the pair to the
 * smaller, phase preserved (GB 2365247 A): true chroma pairs have
 * equal magnitudes and pass unchanged, so only asymmetric (luma)
 * energy is reduced. With a trained LUT, each pair gets a per-bin
 * gain looked up from its symmetry ratio (US 7,872,689).
 *
 * Returns the tile's chroma confidence: the kept output energy's mean
 * pair-symmetry ratio, 0 when nothing was kept. */
static float apply_filter(const comp_transform2d_t *t,
                          const fftwf_complex *in, fftwf_complex *out)
{
    const float *tsq = t->threshold_sq;
    float conf_num = 0.0f, conf_den = 0.0f;

    memset(out, 0, sizeof(fftwf_complex) * YCOMPLEX * XCOMPLEX);

    for (int y = 0; y < YTILE; y++) {
        const int y_ref = ((YTILE / 2) + YTILE - y) % YTILE;
        const fftwf_complex *bi = in + y * XCOMPLEX;
        const fftwf_complex *bi_ref = in + y_ref * XCOMPLEX;
        fftwf_complex *bo = out + y * XCOMPLEX;
        fftwf_complex *bo_ref = out + y_ref * XCOMPLEX;

        /* horizontal frequencies that might be chroma: 0.5fsc to 1.5fsc */
        for (int x = XTILE / 8; x <= XTILE / 4; x++) {
            const int x_ref = (XTILE / 2) - x;
            const float threshold_sq = *tsq++;

            if (x == x_ref && y == y_ref) {
                /* the bin is its own reflection: it is a carrier */
                bo[x][0] = bi[x][0];
                bo[x][1] = bi[x][1];
                conf_num += bi[x][0] * bi[x][0] + bi[x][1] * bi[x][1];
                conf_den += bi[x][0] * bi[x][0] + bi[x][1] * bi[x][1];
                continue;
            }

            const float m_in_sq = bi[x][0] * bi[x][0] + bi[x][1] * bi[x][1];
            const float m_ref_sq = bi_ref[x_ref][0] * bi_ref[x_ref][0]
                                 + bi_ref[x_ref][1] * bi_ref[x_ref][1];
            const float lo = m_in_sq < m_ref_sq ? m_in_sq : m_ref_sq;
            const float hi = m_in_sq < m_ref_sq ? m_ref_sq : m_in_sq;
            const float r = hi > 0.0f ? lo / hi : 1.0f;

            /* LF-luma evidence at the pair's baseband difference
             * frequency k - c (both members map to one stored bin by
             * conjugate symmetry); no LF partner marks cross-color.
             * U and V carriers are the two fixed points of the pair
             * reflection (y = YTILE/4 and 3*YTILE/4), so take the
             * stronger of the two difference positions. */
            float g_e = 1.0f;
            if (t->evidence > 0.0f) {
                const fftwf_complex *lf1 =
                    in + (((YTILE / 4) - y + YTILE) % YTILE) * XCOMPLEX
                       + (XTILE / 4) - x;
                const fftwf_complex *lf2 =
                    in + ((3 * (YTILE / 4) - y + YTILE) % YTILE) * XCOMPLEX
                       + (XTILE / 4) - x;
                const float e1 = (*lf1)[0] * (*lf1)[0] + (*lf1)[1] * (*lf1)[1];
                const float e2 = (*lf2)[0] * (*lf2)[0] + (*lf2)[1] * (*lf2)[1];
                const float e = e1 > e2 ? e1 : e2;
                const float den = e + t->evidence * hi;
                if (den > 0.0f)
                    g_e = e / den;
            }

            if (t->use_lut) {
                const int bin = (int)(tsq - t->threshold_sq) - 1;
                const float g = lut_gain(t->lut[bin], lo, hi) * g_e;
                bo[x][0] = bi[x][0] * g;
                bo[x][1] = bi[x][1] * g;
                bo_ref[x_ref][0] = bi_ref[x_ref][0] * g;
                bo_ref[x_ref][1] = bi_ref[x_ref][1] * g;
                const float e = g * g * (m_in_sq + m_ref_sq);
                conf_num += e * r;
                conf_den += e;
                continue;
            }

            if (t->level) {
                float f_in = g_e, f_ref = g_e;
                if (m_in_sq > m_ref_sq)
                    f_in *= sqrtf(m_ref_sq / m_in_sq);
                else if (m_ref_sq > m_in_sq)
                    f_ref *= sqrtf(m_in_sq / m_ref_sq);
                bo[x][0] = bi[x][0] * f_in;
                bo[x][1] = bi[x][1] * f_in;
                bo_ref[x_ref][0] = bi_ref[x_ref][0] * f_ref;
                bo_ref[x_ref][1] = bi_ref[x_ref][1] * f_ref;
                conf_num += 2.0f * lo * g_e * g_e * r;
                conf_den += 2.0f * lo * g_e * g_e;
                continue;
            }

            if (m_in_sq < m_ref_sq * threshold_sq ||
                m_ref_sq < m_in_sq * threshold_sq)
                continue;  /* asymmetric: probably not chroma */

            bo[x][0] = bi[x][0] * g_e;
            bo[x][1] = bi[x][1] * g_e;
            bo_ref[x_ref][0] = bi_ref[x_ref][0] * g_e;
            bo_ref[x_ref][1] = bi_ref[x_ref][1] * g_e;
            conf_num += (m_in_sq + m_ref_sq) * g_e * g_e * r;
            conf_den += (m_in_sq + m_ref_sq) * g_e * g_e;
        }
    }

    return conf_den > 0.0f ? conf_num / conf_den : 0.0f;
}

void comp_transform2d_field(const comp_transform2d_t *t,
                            const uint16_t *comp, ptrdiff_t comp_stride,
                            int width, int rows,
                            float *chroma, ptrdiff_t chroma_stride,
                            float *conf, ptrdiff_t conf_stride)
{
    float real[YTILE * XTILE];
    fftwf_complex cplx_in[YCOMPLEX * XCOMPLEX];
    fftwf_complex cplx_out[YCOMPLEX * XCOMPLEX];

    for (int r = 0; r < rows; r++) {
        memset(chroma + r * chroma_stride, 0, sizeof(float) * width);
        if (conf)
            memset(conf + r * conf_stride, 0, sizeof(float) * width);
    }

    for (int tile_y = -YTILE / 2; tile_y < rows; tile_y += YTILE / 2) {
        const int start_y = tile_y < 0 ? -tile_y : 0;
        const int end_y = rows - tile_y < YTILE ? rows - tile_y : YTILE;

        for (int tile_x = -XTILE / 2; tile_x < width; tile_x += XTILE / 2) {
            const int start_x = tile_x < 0 ? -tile_x : 0;
            const int end_x = width - tile_x < XTILE ? width - tile_x : XTILE;

            /* windowed forward FFT; samples beyond the raster are black */
            for (int y = 0; y < YTILE; y++) {
                const int row_valid = y >= start_y && y < end_y;
                const uint16_t *b = row_valid ? comp + (tile_y + y) * comp_stride : NULL;
                for (int x = 0; x < XTILE; x++) {
                    const float v = (!row_valid || x < start_x || x >= end_x)
                                    ? LEVEL_BLACK : (float)b[tile_x + x];
                    real[y * XTILE + x] = v * t->window[y][x];
                }
            }
            fftwf_execute_dft_r2c(t->forward, real, cplx_in);

            const float w = apply_filter(t, cplx_in, cplx_out);

            /* inverse FFT; overlap-add the active part, normalized */
            fftwf_execute_dft_c2r(t->inverse, cplx_out, real);
            for (int y = start_y; y < end_y; y++) {
                float *b = chroma + (tile_y + y) * chroma_stride;
                for (int x = start_x; x < end_x; x++)
                    b[tile_x + x] += real[y * XTILE + x] / (YTILE * XTILE);
            }

            /* the half-overlapped windows sum to unity, so accumulating
             * confidence with the same weights yields a [0,1] map */
            if (conf) {
                for (int y = start_y; y < end_y; y++) {
                    float *b = conf + (tile_y + y) * conf_stride;
                    for (int x = start_x; x < end_x; x++)
                        b[tile_x + x] += w * t->window[y][x];
                }
            }
        }
    }
}

File: src/transform2d.zig
const std = @import("std");
const constants = @import("constants.zig");
const fft = @import("fft.zig");

pub const Transform2d = struct {
    pub const Error = error{InvalidThreshold};
    const width = constants.Transform2d.tile_width;
    const height = constants.Transform2d.tile_height;
    const bins = constants.Transform2d.threshold_bins;

    level: bool,
    use_lut: bool,
    evidence: f32,
    window: [height][width]f32,
    threshold_squared: [bins]f32,
    lut: [bins][constants.lut_knots]f32,

    pub fn init(threshold: f64, level: bool, evidence: f64) Error!Transform2d {
        if (!(threshold > 0.0 and threshold <= 1.0) or evidence < 0.0) return error.InvalidThreshold;
        var result = Transform2d{
            .level = level,
            .use_lut = false,
            .evidence = @floatCast(evidence),
            .window = undefined,
            .threshold_squared = undefined,
            .lut = undefined,
        };
        const threshold_sq: f32 = @floatCast(threshold * threshold);
        for (&result.threshold_squared) |*value| value.* = threshold_sq;
        for (0..height) |y| {
            const wy = raisedCosine(y, height);
            for (0..width) |x| result.window[y][x] = wy * raisedCosine(x, width);
        }
        return result;
    }

    pub fn setThresholds(self: *Transform2d, values: []const f64) Error!void {
        if (values.len != bins) return error.InvalidThreshold;
        for (values, 0..) |value, i| {
            if (!(value > 0.0 and value <= 1.0)) return error.InvalidThreshold;
            self.threshold_squared[i] = @floatCast(value * value);
        }
        self.use_lut = false;
    }

    pub fn setLut(self: *Transform2d, values: []const f64) Error!void {
        if (values.len != bins * constants.lut_knots) return error.InvalidThreshold;
        for (0..bins) |bin| {
            for (0..constants.lut_knots) |knot| {
                const value = values[bin * constants.lut_knots + knot];
                if (!(value >= 0.0 and value <= 1.0)) return error.InvalidThreshold;
                self.lut[bin][knot] = @floatCast(value);
            }
        }
        self.use_lut = true;
    }

    pub fn field(self: *const Transform2d, composite: []const u16, composite_stride: usize, raster_width: usize, rows: usize, chroma: []f32, chroma_stride: usize, confidence: ?[]f32, confidence_stride: usize) void {
        std.debug.assert(raster_width <= constants.Standard.pal.activeWidth());
        std.debug.assert(chroma.len >= chroma_stride * rows);
        for (0..rows) |row| {
            @memset(chroma[row * chroma_stride ..][0..raster_width], 0.0);
            if (confidence) |map| @memset(map[row * confidence_stride ..][0..raster_width], 0.0);
        }

        var real: [width * height]f32 = undefined;
        var input: [width * height]fft.Complex = undefined;
        var output: [width * height]fft.Complex = undefined;
        var tile_y: isize = -@as(isize, @intCast(height / 2));
        while (tile_y < @as(isize, @intCast(rows))) : (tile_y += @as(isize, @intCast(height / 2))) {
            const start_y: usize = if (tile_y < 0) @intCast(-tile_y) else 0;
            const remaining_y = @as(isize, @intCast(rows)) - tile_y;
            const end_y: usize = if (remaining_y < @as(isize, @intCast(height))) @intCast(remaining_y) else height;
            var tile_x: isize = -@as(isize, @intCast(width / 2));
            while (tile_x < @as(isize, @intCast(raster_width))) : (tile_x += @as(isize, @intCast(width / 2))) {
                const start_x: usize = if (tile_x < 0) @intCast(-tile_x) else 0;
                const remaining_x = @as(isize, @intCast(raster_width)) - tile_x;
                const end_x: usize = if (remaining_x < @as(isize, @intCast(width))) @intCast(remaining_x) else width;

                for (0..height) |y| {
                    const valid_row = y >= start_y and y < end_y;
                    const source_row: []const u16 = if (valid_row) composite[@as(usize, @intCast(tile_y + @as(isize, @intCast(y)))) * composite_stride ..] else &[_]u16{};
                    const x0 = if (valid_row) start_x else width;
                    const x1 = if (valid_row) end_x else width;
                    const window_row = self.window[y];
                    const destination = real[y * width ..][0..width];
                    for (0..x0) |x| destination[x] = 16384.0 * window_row[x];
                    for (x0..x1) |x| destination[x] = @as(f32, @floatFromInt(source_row[@as(usize, @intCast(tile_x + @as(isize, @intCast(x))))])) * window_row[x];
                    for (x1..width) |x| destination[x] = 16384.0 * window_row[x];
                }

                fft.forward2d(&real, &input, width, height);
                const tile_confidence = self.applyFilter(&input, &output);
                fft.inverse2d(&output, &real, width, height);
                for (start_y..end_y) |y| {
                    const destination = chroma[@as(usize, @intCast(tile_y + @as(isize, @intCast(y)))) * chroma_stride ..];
                    for (start_x..end_x) |x| destination[@as(usize, @intCast(tile_x + @as(isize, @intCast(x))))] += real[y * width + x];
                    if (confidence) |map| {
                        const confidence_row = map[@as(usize, @intCast(tile_y + @as(isize, @intCast(y)))) * confidence_stride ..];
                        for (start_x..end_x) |x| confidence_row[@as(usize, @intCast(tile_x + @as(isize, @intCast(x))))] += tile_confidence * self.window[y][x];
                    }
                }
            }
        }
    }

    fn applyFilter(self: *const Transform2d, input: []const fft.Complex, output: []fft.Complex) f32 {
        @memset(output, .{});
        var confidence_numerator: f32 = 0.0;
        var confidence_denominator: f32 = 0.0;
        var bin: usize = 0;
        for (0..height) |y| {
            const reflected_y = (height / 2 + height - y) % height;
            for (width / 8..width / 4 + 1) |x| {
                const reflected_x = width / 2 - x;
                const source = input[y * width + x];
                const reflected = input[reflected_y * width + reflected_x];
                if (x == reflected_x and y == reflected_y) {
                    output[y * width + x] = source;
                    const energy = source.magnitudeSquared();
                    confidence_numerator += energy;
                    confidence_denominator += energy;
                    bin += 1;
                    continue;
                }

                const input_energy = source.magnitudeSquared();
                const reflected_energy = reflected.magnitudeSquared();
                const low = @min(input_energy, reflected_energy);
                const high = @max(input_energy, reflected_energy);
                const ratio = if (high > 0.0) low / high else 1.0;
                var evidence_gain: f32 = 1.0;
                if (self.evidence > 0.0) {
                    const lf_y1 = (height / 4 + height - y) % height;
                    const lf_y2 = (3 * (height / 4) + height - y) % height;
                    const lf_x = width / 4 - x;
                    const e1 = input[lf_y1 * width + lf_x].magnitudeSquared();
                    const e2 = input[lf_y2 * width + lf_x].magnitudeSquared();
                    const evidence_energy = @max(e1, e2);
                    const denominator = evidence_energy + self.evidence * high;
                    if (denominator > 0.0) evidence_gain = evidence_energy / denominator;
                }

                if (self.use_lut) {
                    const gain = lutGain(self.lut[bin][0..], low, high) * evidence_gain;
                    output[y * width + x] = source.scale(gain);
                    output[reflected_y * width + reflected_x] = reflected.scale(gain);
                    const energy = gain * gain * (input_energy + reflected_energy);
                    confidence_numerator += energy * ratio;
                    confidence_denominator += energy;
                } else if (self.level) {
                    var source_gain = evidence_gain;
                    var reflected_gain = evidence_gain;
                    if (input_energy > reflected_energy and input_energy > 0.0) source_gain *= @sqrt(reflected_energy / input_energy);
                    if (reflected_energy > input_energy and reflected_energy > 0.0) reflected_gain *= @sqrt(input_energy / reflected_energy);
                    output[y * width + x] = source.scale(source_gain);
                    output[reflected_y * width + reflected_x] = reflected.scale(reflected_gain);
                    const energy = 2.0 * low * evidence_gain * evidence_gain;
                    confidence_numerator += energy * ratio;
                    confidence_denominator += energy;
                } else if (!(input_energy < reflected_energy * self.threshold_squared[bin] or reflected_energy < input_energy * self.threshold_squared[bin])) {
                    output[y * width + x] = source.scale(evidence_gain);
                    output[reflected_y * width + reflected_x] = reflected.scale(evidence_gain);
                    const energy = (input_energy + reflected_energy) * evidence_gain * evidence_gain;
                    confidence_numerator += energy * ratio;
                    confidence_denominator += energy;
                }
                bin += 1;
            }
        }
        return if (confidence_denominator > 0.0) confidence_numerator / confidence_denominator else 0.0;
    }
};

fn raisedCosine(element: usize, limit: usize) f32 {
    const angle = 2.0 * std.math.pi * (@as(f64, @floatFromInt(element)) + 0.5) / @as(f64, @floatFromInt(limit));
    return @floatCast(0.5 - 0.5 * @cos(angle));
}

fn lutGain(row: []const f32, low: f32, high: f32) f32 {
    const ratio = if (high > 0.0) low / high else 1.0;
    const position = ratio * @as(f32, @floatFromInt(constants.lut_knots - 1));
    var index: usize = @intFromFloat(position);
    if (index > constants.lut_knots - 2) index = constants.lut_knots - 2;
    return row[index] + (row[index + 1] - row[index]) * (position - @as(f32, @floatFromInt(index)));
}

test "2D transform rejects invalid thresholds" {
    try std.testing.expectError(error.InvalidThreshold, Transform2d.init(0.0, false, 0.0));
    try std.testing.expectError(error.InvalidThreshold, Transform2d.init(1.1, false, 0.0));
    try std.testing.expectError(error.InvalidThreshold, Transform2d.init(0.4, false, -1.0));
}

test "2D transform has unity window centre" {
    const transform = try Transform2d.init(0.4, false, 0.0);
    try std.testing.expect(transform.window[8][16] > 0.99);
}
